### ISSUE TITLE
503s to one of two subsets despite all pods being healthy

### ISSUE THREAD

**Issue by klarose** [istio/istio#20367]
**Bug description**
For a period of time, all requests to one of two subsets in our cluster received a 503 with a response flag UH, indicating that no upstream hosts are healthy. However, there were clearly two healthy upstream hosts which should have mapped to that subset.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Requests should have been forwarded to one of the two hosts which was up.

**Steps to reproduce the bug**
This has only happened once that I have seen, so it may not be very reproducible.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

istio:
```
client version: 1.4.2
control plane version: 1.4.3
data plane version: 1.4.2 (7 proxies), 1.4.3 (19 proxies)
```
Note that we were running 1.4.2 when the issue occurred; we had not yet upgraded to 1.4.3.
 
kubernetes:

```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-10T03:03:57Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.11-gke.15", GitCommit:"bfafff1bd7f2bdc7a379e78d03902f32227cef93", GitTreeState:"clean", BuildDate:"2019-11-25T19:37:48Z", GoVersion:"go1.12.11b4", Compiler:"gc", Platform:"linux/amd64"}
```

helm:

```
version.BuildInfo{Version:"v3.0.2", GitCommit:"19e47ee3283ae98139d98460de796c1be1e3975f", GitTreeState:"clean", GoVersion:"go1.13.5"}
```

**How was Istio installed?**
helm template, followed by kustomize.

**Environment where bug was observed (cloud vendor, OS, etc)**

GKE

**Extra Info**

I reported this issue here: https://discuss.istio.io/t/503s-to-one-of-two-subsets-despite-all-pods-being-healthy/4972

***Setup***

* Istio 1.4.2
* Two subsets (`stable` and `master`).
* One k8s service covering the two subsets.
* One virtual service: We route with cookie to one (`master`). If cookie not set, go to other (`stable`). 2 pods in each subset. 
* We have two pilots running.
* Roughly 127 total pods (via `kubectl get pods --all-namespaces | grep -v NAMESPACE | wc -l`)
* Roughly 93 services
* Roughly 60 virtual services

***Symptom***
All requests to `stable` receive a 503 with response_flags UH. Prometheus stats (envoy_cluster_membership_healthy) show no entries active in the cluster corresponding to subset `stable`. The cluster corresponding to the entire service shows 4 entries. EDS retrieved from istioctl to one of the ingresses shows no rows for `stable`, but rows for `master`.

***Recovery***
 Restarting one of the two running pilots resolved the problem. We first tried restarting one of the ingress gateways, which did not fix the problem. The pilot I restarted had more in the logs (various EDS/ADS push events) than the other, which is why I chose it.

***Timeline***

* At around 16:24 GMT the node which was running both the pilot I restarted and the two `stable` pods rebooted.
* When the node came back up the `stable` pods were started at around 16:25.
* From the time the node rebooted until 16:44, both ingress gateways reported the `stable` cluster as having no members. This is quite a bit longer than expected. The pods were definitely up during this time, as the cluster containing all pods for the service showed 4 members in prometheus.
* From 16:44 GMT onward, one of the two ingress gateways was reporting the `stable` cluster as being down. Restarting that gateway fixed the problem. This may or may not be related.
* At about 17:30 GMT, one of the ingress gateways reported the `stable` cluster as having no members in the envoy_cluster_membership_healthy stat
* At about 17:36 GMT the other ingress gateway reported the same.
* The system stayed in this state until I restarted pilot.

***Other Notes***

Endpoints before restarting pilot.

```
istioctl proxy-config endpoints -n istio-system istio-ingressgateway-78b57d75d-kgfp2 | grep portal
10.16.0.20:5000        HEALTHY     OK                outbound_.5000_._.portal.portal.svc.cluster.local
10.16.0.20:5000        HEALTHY     OK                outbound|5000||portal.portal.svc.cluster.local
10.16.0.30:5000        HEALTHY     OK                outbound_.5000_._.portal.portal.svc.cluster.local
10.16.0.30:5000        HEALTHY     OK                outbound|5000||portal.portal.svc.cluster.local
10.16.2.31:5000        HEALTHY     OK                outbound_.5000_._.portal.portal.svc.cluster.local
10.16.2.31:5000        HEALTHY     OK                outbound_.5000_.master_.portal.portal.svc.cluster.local
10.16.2.31:5000        HEALTHY     OK                outbound|5000|master|portal.portal.svc.cluster.local
10.16.2.31:5000        HEALTHY     OK                outbound|5000||portal.portal.svc.cluster.local
10.16.4.27:5000        HEALTHY     OK                outbound_.5000_._.portal.portal.svc.cluster.local
10.16.4.27:5000        HEALTHY     OK                outbound_.5000_.master_.portal.portal.svc.cluster.local
10.16.4.27:5000        HEALTHY     OK                outbound|5000|master|portal.portal.svc.cluster.local
10.16.4.27:5000        HEALTHY     OK                outbound|5000||portal.portal.svc.cluster.local
```

Endpoints after restarting pilot. Note how stable is now present, but we still have the same set of IPs:

```
istioctl proxy-config endpoints -n istio-system istio-ingressgateway-78b57d75d-kgfp2 | grep portal
10.16.0.20:5000        HEALTHY     OK                outbound_.5000_._.portal.portal.svc.cluster.local
10.16.0.20:5000        HEALTHY     OK                outbound_.5000_.stable_.portal.portal.svc.cluster.local
10.16.0.20:5000        HEALTHY     OK                outbound|5000|stable|portal.portal.svc.cluster.local
10.16.0.20:5000        HEALTHY     OK                outbound|5000||portal.portal.svc.cluster.local
10.16.0.30:5000        HEALTHY     OK                outbound_.5000_._.portal.portal.svc.cluster.local
10.16.0.30:5000        HEALTHY     OK                outbound_.5000_.stable_.portal.portal.svc.cluster.local
10.16.0.30:5000        HEALTHY     OK                outbound|5000|stable|portal.portal.svc.cluster.local
10.16.0.30:5000        HEALTHY     OK                outbound|5000||portal.portal.svc.cluster.local
10.16.2.31:5000        HEALTHY     OK                outbound_.5000_._.portal.portal.svc.cluster.local
10.16.2.31:5000        HEALTHY     OK                outbound_.5000_.master_.portal.portal.svc.cluster.local
10.16.2.31:5000        HEALTHY     OK                outbound|5000|master|portal.portal.svc.cluster.local
10.16.2.31:5000        HEALTHY     OK                outbound|5000||portal.portal.svc.cluster.local
10.16.4.27:5000        HEALTHY     OK                outbound_.5000_._.portal.portal.svc.cluster.local
10.16.4.27:5000        HEALTHY     OK                outbound_.5000_.master_.portal.portal.svc.cluster.local
10.16.4.27:5000        HEALTHY     OK                outbound|5000|master|portal.portal.svc.cluster.local
10.16.4.27:5000        HEALTHY     OK                outbound|5000||portal.portal.svc.cluster.local
```

The destination rule we are using to build the subsets:

```
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: portal
  namespace: portal
spec:
  host: portal
  subsets:
    - name: stable
      labels:
        ver: stable
    - name: master
      labels:
        ver: master
  trafficPolicy:
    connectionPool:
      tcp:
        connectTimeout: 1500ms
```

**Comment by TomasKohout** [comment]
This is fixed in 1.4.5. @klarose could you close this issue please? :-)

**Comment by klarose** [comment]
Yes. 1.4.5 appears to have fixed it. Thanks.
